Provide a Promise based API

jonathansamines opened the pull request:
Have you considered providing an Promise based API? that would be useful for people who is working with async programming through promises.  Here is a pull request providing this dual api.

andreareginato commented:
Hi @jonathansamines. This pull request looks really interesting. One question. Did you have any time on checking out tests and thinking about some documentation that could vary or be added to promises usage? I think it would be great for developers to have more docs to check out

jonathansamines commented:
Hi @andreareginato i just was thinking about the test but as i can understand from your test, making a test to the promised-api, should lead to a all test duplication no? because right now all async all are made by using the beforeEach blocks. I´m fine doing that, but did you know if there is a better way to achieve this task? 

On the documentation side, i will work on that this night. I think that most of public methods that changed, are really easy to convert to a promised based ones on the docs.

andreareginato commented:
This duplication doesn't sound good. I don't have an idea of how to make
this work right, but I think some researches need to be done.
Great for the documentation.

On Tue, Sep 29, 2015 at 5:45 PM, Jonathan Samines <notifications@github.com>
wrote:

> Hi @andreareginato <https://github.com/andreareginato> i just was
> thinking about the test but as i can understand from your test, making a
> test to the promised-api, should lead to a all test duplication no? because
> right now all async all are made by using the beforeEach blocks. I´m fine
> doing that, but did you know if there is a better way to achieve this task?
>
> On the documentation side, i will work on that this night. I think that
> most of public methods that changed, are really easy to convert to a
> promised based ones on the docs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/andreareginato/simple-oauth2/pull/46#issuecomment-144098913>
> .
>



-- 
Andrea Reginato
Lelylan | reThink your house
http://lelylan.com

jonathansamines commented:
Hi! finally i finished and refactored all test to support all the promise based api. I also updated the docs to provide some examples of promise usage. Please let me know if it is ok all changes made.

andreareginato commented:
Looks great to me. Thanks @jonathansamines. 

I would love to ask you one thing. As I'm out of time, would you like to be added as admin and handle the merge. It looks like there are some conflicts needs to be resolved. 

Your help would be really appreciated.

jonathansamines commented:
I no have problem with being added as an admin. Only give me the permissions, and i´ll handle the merge.

andreareginato commented:
You are in. Thanks a lot. 

I would also love to ask you one more thing, but preferably in a private conversation.
I'll be back to you next days.

weiwanning commented on package.json:
```diff
@@ -17,9 +17,10 @@
   },
   "devDependencies": {
-    "should": "~1.2.1",
+    "bluebird": "^2.10.1",
```
Any reasons to add bluebird in devDependencies?

jonathansamines replied:
In fact, it´s supposed to be a dependency instead of a devDependency and the reason is that i use the "dual api" support from this lib (nodeify), to add support to promises to this lib. I´ll move this dependency to the correct place. A new release number hasn´t been generated yet, so we can still fix it.
